src/v/datalake/translation/partition_translator.h

```diff
@@ -0,0 +1,166 @@
+ * to low level iceberg translator that converts the data.
+ *
+ * The translator works in tandem with the datalake coordinator responsible for
+ * this ntp. A successful local iceberg translation of a data range is followed
+ * by that state checkpointing with the coordinator. A coordinator tracks the
+ * latest translated offset for a given ntp (along with other metadata) which is
+ * also then synchronized with the local translation stm that enforces
+ * max_collectible_offset across all replicas. The translator tries to work
+ * with the stm state as much as possible (to avoid round trips to the
+ * coordinator)
+ *
+ * It is possible that the translation stm can get out of sync with what is
+ * on the coordinator, so it is reconciled as needed. It is ok to work on a
+ * stale stm state as the eventual checkpoint request with the coordinator
+ * fenced based on the translator term.
+ *
+ * The logic for this translator is like this..
+ *
+ * while (!aborted && !term_changed):
+ *    sleep(interval)
+ *    reconcile_with_coordinator_if_needed()
+ *    md = translate_newly_arrived_data_since_last_checkpoint()
+ *    checkpoint_with_coordinator(md)
+ *    sync_stm_with_coordinator(md)
+ */
+
+class partition_translator {
+public:
+    explicit partition_translator(
+      ss::lw_shared_ptr<cluster::partition> partition,
+      ss::sharded<coordinator::frontend>* frontend,
+      ss::sharded<features::feature_table>* features,
+      std::chrono::milliseconds translation_interval,
+      ss::scheduling_group sg,
+      size_t reader_max_bytes,
+      ssx::semaphore& parallel_translations);
+
+    ss::future<> stop();
+
+    std::chrono::milliseconds translation_interval() const;
+    void reset_translation_interval(std::chrono::milliseconds new_base);
+
+private:
+    struct backoff {
+        static constexpr std::chrono::milliseconds initial_sleep{300};
+        explicit backoff(std::chrono::milliseconds max)
+          : max_sleep(max) {}
+
+        std::chrono::milliseconds next() {
+            current_sleep = std::min(
+              std::max(initial_sleep, current_sleep * 2), max_sleep);
+            return current_sleep;
+        }
+
+        void reset() { current_sleep = std::chrono::milliseconds{1}; }
+
+        template<
+          typename Func,
+          typename ShouldRetry,
+          typename FuncRet = std::invoke_result_t<Func>>
+        requires std::
+          predicate<ShouldRetry, typename ss::futurize<FuncRet>::value_type>
+          ss::futurize_t<FuncRet>
+          retry(int num_attempts, Func&& f, ShouldRetry&& should_retry) {
+            reset();
+            while (num_attempts-- > 0) {
+                auto result = co_await ss::futurize_invoke(f);
+                if (num_attempts > 0 && should_retry(result)) {
+                    co_await ss::sleep(next());
+                    continue;
+                }
+                co_return result;
+            }
+            __builtin_unreachable();
```

Comment:
Is this really unreachable? It seems like if you are retrying something it's valid to exhaust retries - should we really abort the process? At least `vassert` for better logging...

Reply:
I guess the intention is that `should_retry` should return false before `max_attempts` is added. Should we rewrite the loop so that the behavior is the same if someone is a little careless or we have a logic bug in `should_retry`? May just throw an exception.